### PR TITLE
fix(material/autocomplete): autocomplete panel top is cut off in landscape mode

### PR DESCRIPTION
**Pull request by essjay05**
Fixes a bug in the Angular Material Autocomplete component where the autocomplete panel listbox top is cut off by the device screen when being viewed in landscape mode. This is because the listbox max-height is greater than the distance between the top of the autocomplete input and the screen top.

Fixes b/284148377

### CODE REVIEW

**Comment by essjay05** [comment]
Googlers see [internal](https://b.corp.google.com/issues/284148377).

Screenshot for bug fix test reference [here](https://screenshot.googleplex.com/9utanYh4obyo7og).

File: src/material/autocomplete/autocomplete.scss [hunk 1]
```diff
@@ -54,3 +54,10 @@ div.mat-mdc-autocomplete-panel {
 mat-autocomplete {
   display: none;
 }
+
+// Prevents autocomplete panel from being cut-off from being viewed in landscape orientation.
+@media all and (orientation: landscape) {
+  div.mat-mdc-autocomplete-panel {
+    max-height: 55vh;
```
[the remark this answers is not in the document]
Comment (essjay05) [reply src/material/autocomplete/autocomplete.scss:61]: Thanks for the feedback @crisbeto! Will look into applying your recommended solution.

**Comment by essjay05** [comment]
Screencast after recommended changes [here](https://screencast.googleplex.com/cast/NjQ1NzY4MzU2ODc1NDY4OHxmMmY2ZThhNC0wMQ)

**Review by crisbeto — APPROVED** [review]

File: src/material/autocomplete/autocomplete-trigger.ts [hunk 2]
```diff
@@ -879,13 +880,29 @@ export class MatAutocompleteTrigger
...
   private _getOverlayPosition(): PositionStrategy {
+    // Set default Overlay Position
     const strategy = this._overlay
       .position()
       .flexibleConnectedTo(this._getConnectedElement())
       .withFlexibleDimensions(false)
       .withPush(false);
 
+    // Check breakpoint if being viewed in HandsetLandscape
+    const isHandsetLandscape = this._breakpointObserver
```
Comment (mmalerba) [inline src/material/autocomplete/autocomplete-trigger.ts:895]: Isn't this a subscription object not a boolean? Also I think we need to unsubscribe / takeUntil(destroyed)

Comment (essjay05) [reply src/material/autocomplete/autocomplete-trigger.ts]: @mmalerba Thanks for the note! I've made the change to unsubscribe in my latest commit. Let me know if it's good to go and I'll make the request to merge.

File: src/material/autocomplete/autocomplete-trigger.ts [hunk 3]
```diff
@@ -879,13 +882,35 @@ export class MatAutocompleteTrigger
...
+
   private _getOverlayPosition(): PositionStrategy {
+    // Set default Overlay Position
     const strategy = this._overlay
       .position()
       .flexibleConnectedTo(this._getConnectedElement())
       .withFlexibleDimensions(false)
       .withPush(false);
 
+    if (!this._handsetLandscapeBreakpointSubscription) {
```
Comment (mmalerba) [inline src/material/autocomplete/autocomplete-trigger.ts:896]: This is kind of a weird place to subscribe (as a side-effect in a getter function) can you move it to `_attachOverlay` right after we create `_viewportSubscription`

File: src/material/autocomplete/autocomplete-trigger.ts [hunk 4]
```diff
@@ -879,13 +882,35 @@ export class MatAutocompleteTrigger
...
+      // BreakpointObserver only returns screen size or isMatched/matches boolean.
+      this._handsetLandscapeBreakpointSubscription = this._breakpointObserver
+        .observe(Breakpoints.HandsetLandscape)
+        .subscribe(result => {
+          // Check breakpoint if being viewed in HandsetLandscape via subscription (if not null)
+          const isHandsetLandscape = result.matches;
+          // Apply HandsetLandscape settings to prevent overlay cutoff in that breakpoint
+          // Fixes b/284148377
+          if (isHandsetLandscape) {
+            strategy.withFlexibleDimensions(true).withGrowAfterOpen(true).withViewportMargin(8);
```
Comment (mmalerba) [inline src/material/autocomplete/autocomplete-trigger.ts:908]: do we need to unset these options if it changes back to portrait?

File: src/material/autocomplete/autocomplete-trigger.ts [hunk 5]
```diff
@@ -789,6 +795,23 @@ export class MatAutocompleteTrigger
...
+          const isHandsetLandscape = result.matches;
+          // Check if result.matches Breakpoints.HandsetLandscape. Apply HandsetLandscape
+          // settings to prevent overlay cutoff in that breakpoint. Fixes b/284148377
+          if (isHandsetLandscape) {
+            this._positionStrategy
+              .withFlexibleDimensions(true)
+              .withGrowAfterOpen(true)
+              .withViewportMargin(8);
+          } else {
+            this._positionStrategy.withGrowAfterOpen(false);
```
Comment (mmalerba) [inline src/material/autocomplete/autocomplete-trigger.ts:812]: You also need to set `withFlexibleDimensions` and `withViewportMargin` back to their original values right?

**Review by mmalerba — APPROVED** [review]

File: src/material/autocomplete/autocomplete-trigger.ts (hunk 4 above)
Comment (essjay05) [reply src/material/autocomplete/autocomplete-trigger.ts]: Updated to set to previous default settings if isHandsetLandscape is false.

File: src/material/autocomplete/autocomplete-trigger.ts (hunk 5 above)
Comment (essjay05) [reply src/material/autocomplete/autocomplete-trigger.ts]: Updated!

**Pull request merged**